### DISCUSSION TITLE
Fatture - Campo 2.1.1.1 - TipoDocumento

### DISCUSSION THREAD

**Discussion by g-scalvini** [fattureincloud/api#398] [Q&A] [answered]
Buongiorno a tutti,
all'interno di un'integrazione custom con Fatture in Cloud abbiamo la necessità di specificare, nella chiamata di creazione di una fattura (POST /c/{company_id}/issued_documents), il campo 2.1.1.1 - TipoDocumento.

Tale campo è modificabile da interfaccia FiC nella sezione "Attributi avanzati" del blocco Fatturazione elettronica mentre all'interno del SDK o della documentazione API non riesco a trovare riferimenti al campo in questione.
Mi aspettavo di trovarlo nella classe FattureInCloud\Model\IssuedDocumentEiData dell PHP SDK ma vedo che non è presente.

Potreste indicarmi se tale campo è esposto nella rotta API oppure se ne è prevista la futura implementazione?

Grazie

**Comment by valmoz** [comment, marked as answer]
Buongiorno @g-scalvini 
per impostare questo campo può utilizzare il campo ei_raw, come descritto in [questa guida](https://developers.fattureincloud.it/docs/guides/e-invoice-xml-customisation).

La guida ha anche degli esempi basati sulle SDK come ulteriore supporto.

Restiamo a disposizione per ulteriori domande.

Grazie
Mauro

**Reply by g-scalvini** [reply]
Buongiorno Mauro,
la ringrazio per il celere riscontro.

Ho effettuato le modifiche come da guida indicata e sono riuscito a valorizzare correttamente il campo "TipoDocumento".

Vorrei segnalare le seguenti anomalie che ho notato:

1. Nella guida, per il PHP SDK viene mostrato l'utilizzo di un array come parametro per il metodo IssuedDocument::setEiRaw() ed effettivamente utilizzando un array la logica funziona correttamente. La documentazione del metodo indica però che il campo dovrebbe essere `@param object|null $ei_raw` ed infatti, utilizzando un array, l'IDE (PHPStorm) segnala il confilitto: "Expected parameter of type 'null|object', 'array[]' provided".
2. Sembra non esserci un controllo sulla struttura dei dati inviati tramite il metodo IssuedDocument::setEiRaw().
Ad esempio, valorizzando erroneamente "DatiGeneraliDocumento" <1.1> come una lista JSON la chiamata di creazione va a buon fine e la fattura appare in FiC, ma provando a visualizzarla o a scaricare l'XML, questo è completamente invalido:
```
Errore interpretazione XML: non well-formed
Indirizzo: https://secure.fattureincloud.it/backend_apiV2/e_invoices/preview?invoice_token=XXX&stylesheet=assosoftware
Riga numero 65, colonna 10:
        <0>Migrate to v2 - Oggetto (visibile)</0>
---------^
```
ed infatti nell'XML sono presenti N tag `<DatiGeneraliDocumento>`:
```
OMISS
<DatiGenerali>
      <DatiGeneraliDocumento>
        <TipoDocumento>TD02</TipoDocumento>
      </DatiGeneraliDocumento>
      <DatiGeneraliDocumento>TD01</DatiGeneraliDocumento>
      <DatiGeneraliDocumento>EUR</DatiGeneraliDocumento>
      <DatiGeneraliDocumento>2024-01-11</DatiGeneraliDocumento>
      <DatiGeneraliDocumento>20230216</DatiGeneraliDocumento>
      <DatiGeneraliDocumento>183.00</DatiGeneraliDocumento>
      <DatiGeneraliDocumento>
        <0>Migrate to v2 - Oggetto (visibile)</0>
      </DatiGeneraliDocumento>
OMISS
```
Giorgio

**Reply by emajo** [reply]
Ciao @g-scalvini, è vero nella documentazione utilizziamo un array per valorizzare il campo ei_raw ma non dovrebbe esserci nessun problema a utilizzare un oggetto StdClass così da non avere errore nell'ide.
Per quanto riguarda la validazione del campo confermo che non è presente, per controllare la validità dell'xml può essere usata in un secondo momento l'apposita chiamata [Verify E-Invoice Xml](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/issued_documents/-document_id-/e_invoice/xml_verify).

Grazie,
Emanuele.

**Reply by g-scalvini** [reply]
Grazie ad entrambi per il supporto.

Buona giornata
Giorgio
